torchao/prototype/mx_formats/custom_cast.py

```diff
@@ -1080,3 +1082,240 @@ def _(uint8_data):
+        n_rows,  # number of rows in the tensor
+        n_cols,  # number of columns in the tensor
+        TILE_SIZE: tl.constexpr,  # tile size as a compile-time constant
+    ):
+        """
+        credit: mostly Claude, some Vasiliy
+        """
+
+        # Get program ID
+        pid_row = tl.program_id(0)
+        pid_col = tl.program_id(1)
+
+        # Calculate starting row and column for this tile
+        start_row = pid_row * TILE_SIZE
+        start_col = pid_col * TILE_SIZE
+
+        # Create offsets for the block
+        row_offsets = tl.arange(0, TILE_SIZE)
+        col_offsets = tl.arange(0, TILE_SIZE)
+
+        # Compute global row/col positions
+        rows = start_row + row_offsets[:, None]  # Convert to 2D for proper broadcasting
+        cols = start_col + col_offsets[None, :]
+
+        # Create masks for out-of-bounds accesses
+        row_mask = rows < n_rows
+        col_mask = cols < n_cols
+        mask = row_mask & col_mask
+
+        # Compute memory offsets for row-major layout (rows, cols)
+        row_major_offsets = (rows * n_cols + cols).to(tl.int32)
+
+        # Compute memory offsets for column-major layout (cols, rows)
+        col_major_offsets = (cols * n_rows + rows).to(tl.int32)
+
+        # Load the entire block in a single operation
+        x_block = tl.load(x_ptr + row_major_offsets, mask=mask)
+
+        # ----------------------------------------------------
+        # Row-wise normalization
+        # ----------------------------------------------------
+        # Calculate the absolute values of elements in the block
+        x_block_abs = tl.abs(x_block)
+
+        # Find the maximum absolute value for each row
+        row_scale, row_scale_e8m0 = _triton_calculate_scale(x_block_abs, axis=1)
+
+        # Normalize each row by its maximum absolute value
+        # Broadcasting row_scale to match x_block's shape
+        row_normalized = x_block / row_scale[:, None]
+
+        # quant to float8
+        # TODO(this PR): clamp?
+        row_normalized = row_normalized.to(tl.float8e4nv)
+
+        # ----------------------------------------------------
+        # Column-wise normalization
+        # ----------------------------------------------------
+        # Find the maximum absolute value for each column
+        col_scale, col_scale_e8m0 = _triton_calculate_scale(x_block_abs, axis=0)
+
+        # Normalize each column by its maximum absolute value
+        # Broadcasting col_scale to match x_block's shape
+        col_normalized = x_block / col_scale[None, :]
+
+        # quant to float8
+        # TODO(this PR): clamp?
+        col_normalized = col_normalized.to(tl.float8e4nv)
+
+        # Store the row-normalized result in row-major format
+        tl.store(output_row_major_ptr + row_major_offsets, row_normalized, mask=mask)
+
+        # Store the column-normalized result in column-major format
+        tl.store(output_col_major_ptr + col_major_offsets, col_normalized, mask=mask)
```

Comment:
if I comment this line out, I get to 60% of peak mem bandwidth (nearly matching an "add one" kernel). uncoalesced writes maybe...